tower-balance/src/p2c/service.rs

```diff
@@ -230,45 +287,58 @@ where
         // previously-selected `ready_index` if appropriate.
         self.poll_discover()?;
 
-        if let Some(index) = self.ready_index {
-            debug_assert!(index < self.endpoints.len());
-            // Ensure the selected endpoint is still ready.
-            match self.poll_endpoint_index_load(index) {
-                Ok(Async::Ready(_)) => return Ok(Async::Ready(())),
-                Ok(Async::NotReady) => {}
-                Err(e) => {
-                    drop(self.endpoints.swap_remove_index(index));
-                    info!("evicting failed endpoint: {}", e.into());
-                }
-            }
-
-            self.ready_index = None;
-        }
+        self.poll_unready();
+        debug!("ready={}; unready={}", self.ready.len(), self.unready.len());
 
-        let tries = match self.endpoints.len() {
-            0 => return Ok(Async::NotReady),
-            1 => 1,
-            n => n / 2,
-        };
-        for _ in 0..tries {
-            if let Async::Ready(idx) = self.poll_ready_index().map_err(Into::into)? {
-                trace!("ready: {:?}", idx);
-                self.ready_index = Some(idx);
+        if let Some(index) = self.next_ready_index {
```

Comment:
I'm not sure I follow this. Why would `self.next_ready_index` every be `Some` here? Don't we always immediately use an endpoint when we select one?

Reply:
We have no guarantees that poll will only be called once before call.

Reply:
Oh, this is me being stupid. `poll_ready` will be called first, and it is what sets `next_ready_index`, which is used by `call`. Never mind me :sweat_smile: